### PR TITLE
Feat/milestone reviewers management

### PR DESCRIPTION
**Pull request by brunodmsi**
## 📋 Overview

This PR implements a comprehensive milestone reviewer management system that allows community admins to assign dedicated reviewers for project milestones, separate from the existing program reviewer role. Milestone reviewers automatically inherit all program reviewer permissions plus the ability to review project milestones.

**✨ Latest Update:** Added security improvements, type safety enhancements, and UX improvements based on code review feedback.

**Related Backend PR:** [gap-indexer#646](https://github.com/show-karma/gap-indexer/pull/646) - Adds API support for milestone reviewer endpoints

#### 1. Milestone Reviewer Management UI
- **File:** `components/FundingPlatform/QuestionBuilder/ReviewerManagementTab.tsx`
- Added dual-role reviewer management interface with role toggle (Program vs Milestone)
- Separate lists for program reviewers and milestone reviewers
- Unified add/remove functionality with role selection
- Proper permission checks and loading states
- **NEW:** Robust member ID parsing with validation
- **NEW:** Improved error messages with specific status codes

#### 2. Milestone Reviewer Service & Hooks
- **Files:** `services/milestone-reviewers.service.ts`, `hooks/useMilestoneReviewers.ts`
- Created `MilestoneReviewersService` for API interactions with `/v2/programs/:programId/:chainID/milestone-reviewers`
- Built `useMilestoneReviewers()` hook with memoized data and mutations
- Consolidated add/remove operations with optimistic updates
- **NEW:** Uses shared validation utilities for consistency

#### 3. Generic Role Management Component
- **Files:** `components/Generic/RoleManagement/RoleManagementTab.tsx`, `components/Generic/RoleManagement/types.ts`, `components/Generic/RoleManagement/helpers.ts`
- Extracted reusable role management component with TypeScript types
- Supports multiple role types through configuration
- Handles form validation, loading states, and error handling
- **NEW:** Helper utilities for role extraction and labeling
- **NEW:** Enhanced confirmation dialogs showing role badges

#### 4. My Reviews Page Enhancements
- **File:** `app/my-reviews/page.tsx`
- Added "View Milestones" button for milestone reviewers
- Tracks milestone reviewer programs per community
- Passes filtered `programIds` query parameter to milestone report page
- Side-by-side button layout (View Programs + View Milestones)

#### 5. Milestone Report Access Control
- **File:** `components/Pages/Admin/ReportMilestonePage.tsx`
- Implemented permission-based access control for milestone reviewers
- Filters program dropdown to only show authorized programs
- Admins and contract owners bypass filtering (see all programs)
- Validates user permissions before showing page content
- **NEW:** URL validation for programIds query parameters
- **NEW:** Input sanitization to prevent tampering

#### 6. Milestone Review Permission Enforcement
- **Files:** `components/Pages/Admin/MilestonesReview/index.tsx`, `MilestoneCard.tsx`
- Added granular permission checks for milestone verification
- Only milestone reviewers, admins, and contract owners can verify/sync milestones
- **NEW:** Distinguishes between program reviewers and milestone reviewers
- **NEW:** `canVerifyMilestones` flag passed to UI components

#### 1. Shared Validation Utilities
- **File:** `utilities/validators.ts` (NEW - 274 lines)
- Centralized validation logic for consistency
- **Functions:**
  - `validateWalletAddress()` - Ethereum address validation
  - `validateEmail()` - Email format validation
  - `validateTelegram()` - Telegram handle validation
  - `validateProgramId()` - Program ID format validation
  - `validateChainId()` - Chain ID validation (positive integer)
  - `validateProgramIdentifier()` - Combined programId_chainID validation
  - `validateProgramIdentifiers()` - Batch validation with detailed errors
  - `parseReviewerMemberId()` - Robust member ID parsing (role-publicAddress)
  - `sanitizeString()` - XSS prevention
  - `validateReviewerData()` - Complete reviewer data validation

#### 2. URL Parameter Validation
- **File:** `components/Pages/Admin/ReportMilestonePage.tsx`
- Validates `programIds` query parameter format before use
- Filters out invalid program IDs automatically
- Shows user-friendly error toast for invalid inputs
- Prevents URL tampering attacks

#### 3. Hardened String Parsing
- **File:** `components/FundingPlatform/QuestionBuilder/ReviewerManagementTab.tsx`
- Replaced fragile `split("-", 2)` with robust parser
- Validates role type is one of known values ("program" | "milestone")
- Validates wallet address format in member ID
- Graceful error handling with user feedback

#### 1. Removed All `any` Types
- **10 instances removed** across all files
- Replaced with proper TypeScript type guards
- **Error handling:** Uses `instanceof Error` and type checking
- **API errors:** Proper type narrowing for axios errors

#### 2. Type Definitions
- **File:** `components/Generic/RoleManagement/types.ts`
- Added `ReviewerRole` type: `"program" | "milestone"`
- Updated `RoleMember` interface with `role?` property
- Changed `[key: string]: any` to `[key: string]: string | undefined`
- **File:** `components/Generic/RoleManagement/helpers.ts` (NEW)
- `getMemberRole()` - Type-safe role extraction
- `getRoleLabel()` - Full role label ("Program Reviewer" / "Milestone Reviewer")
- `getRoleShortLabel()` - Short role label ("Program" / "Milestone")

#### 3. Service Layer Type Safety
- **Files:** `services/program-reviewers.service.ts`, `services/milestone-reviewers.service.ts`
- Proper error type checking without `any`
- Type guards for API response structures
- Explicit type assertions where needed

#### 1. Improved Confirmation Dialogs
- Shows reviewer name AND role in removal confirmation
- Before: "Are you sure you want to remove John Doe?"
- After: "Are you sure you want to remove John Doe (Program Reviewer)?"
- Provides clearer context when same person has multiple roles

#### 2. Better Error Messages
- **Authentication errors (401/403):** "You don't have permission..."
- **Conflict errors (409):** "This member already exists"
- **Validation errors (400):** Shows specific validation message
- **Not found errors (404):** "Member not found. May have been removed."

#### 3. Enhanced Loading States
- Loading spinners on action buttons during mutations
- Individual member removal shows spinner on that specific row
- Disabled state prevents double-clicks
- Clear visual feedback for all async operations

#### 1. Memoization
- Added `useMemo` for `isLoadingReviewers` computation
- Memoized `members` array construction
- Memoized `roleOptions` and config objects
- Memoized validation functions
- All callbacks use `useCallback` with proper dependencies

#### 2. Efficient Data Structures
- Use `Map` for O(1) program lookups
- Use `Set` for permission checking
- Proper dependency arrays to prevent unnecessary re-renders

#### 1. DRY Principle Applied
- **~100 lines of duplicate validation code removed**
- Both reviewer services now use shared validators
- Consistent validation behavior across the app
- Single source of truth for validation rules

#### 2. Separation of Concerns
- **types.ts:** Type definitions only
- **helpers.ts:** Utility functions
- **RoleManagementTab.tsx:** Component implementation
- Clear boundaries between modules

#### Frontend
- **My Reviews Page** (`/my-reviews`): Now shows milestone-specific actions
- **Reviewer Management Tab**: Completely refactored with role toggle
- **Milestone Report Page** (`/community/[id]/admin/milestones-report`): Added access control + validation
- **Milestone Review Page** (`/community/[id]/admin/milestones-review`): Permission enforcement
- **Generic Components**: New reusable RoleManagementTab component with helpers

#### Backend Integration
- Consumes new `/v2/funding-program-configs/my-reviewer-programs` endpoint (returns both reviewer types)
- Uses `/v2/programs/:programId/:chainID/milestone-reviewers` endpoints (GET, POST, DELETE)
- Expects `isProgramReviewer` and `isMilestoneReviewer` flags in API responses

### Permission Model

**Milestone Reviewer = Program Reviewer + Milestone Access**

Milestone reviewers automatically get:
- ✅ All program reviewer permissions
- ✅ Access to review funding applications
- ✅ Access to `/my-reviews` page
- ✅ Access to program reviewer dashboard
- ✅ **PLUS:** Access to milestone review functionality
- ✅ **PLUS:** "View Milestones" button on their dashboard
- ✅ **PLUS:** Ability to verify/sync project milestones

### Breaking Changes
- [x] **No breaking changes** - This is additive functionality
- Existing program reviewer functionality remains unchanged
- New milestone reviewer role is optional
- All validation is backward compatible

### Performance Impact
- [x] **No performance regression**
- Added memoization to prevent unnecessary re-renders
- Efficient data grouping with Map/Set for O(1) lookups
- React Query caching reduces redundant API calls
- Validation functions are lightweight and memoized

### Security Impact
- [x] **Improved security posture**
- Input validation prevents injection attacks
- URL tampering protection
- Type safety prevents runtime errors
- Graceful error handling prevents information leakage

### Prerequisites
```bash
# Ensure backend is running with milestone reviewer endpoints
cd gap-indexer
yarn dev

# Start frontend
cd gap-app-v2
pnpm dev
```

### 1. Milestone Reviewer Assignment (Admin)

**As a Community Admin:**

1. Navigate to a funding program's question builder
2. Click on the "Reviewer Management" tab
3. Toggle to "Milestone Reviewers" role
4. Add a milestone reviewer:
   - Enter wallet address
   - Fill in name, email, telegram (optional)
   - Click "Add Reviewer"
5. Verify reviewer appears in the milestone reviewers list with "Milestone" badge
6. Remove reviewer and verify deletion confirmation shows role
7. Try invalid inputs (bad email, invalid wallet) - should see validation errors

**Expected:**
- ✅ Milestone reviewers list updates immediately
- ✅ Toast notifications for add/remove actions
- ✅ Form clears after successful add
- ✅ Toggle switches between program/milestone lists
- ✅ Validation prevents invalid data
- ✅ Confirmation dialog shows "Remove [Name] (Milestone Reviewer)?"

### 2. My Reviews Page (Milestone Reviewer)

**As a Milestone Reviewer:**

1. Connect wallet assigned as milestone reviewer
2. Navigate to `/my-reviews`
3. Verify you see:
   - "View Programs" button (primary)
   - "View Milestones" button (secondary with eye icon)
4. Click "View Milestones"
5. Verify redirect to `/community/{slug}/admin/milestones-report?programIds={ids}`
6. Check URL contains correct programIds

**Expected:**
- ✅ Both buttons visible for milestone reviewers
- ✅ Correct community grouping
- ✅ Proper program ID filtering in URL
- ✅ Clean button styling and layout

### 3. Milestone Report Access Control & Validation

**As a Milestone Reviewer:**

1. Access milestone report via "View Milestones" button
2. Verify program dropdown only shows your assigned programs
3. Try manually changing `programIds` query param to:
   - Invalid format (e.g., `?programIds=abc,xyz`)
   - Unauthorized program ID
   - Malformed ID (missing underscore)
4. Verify:
   - Invalid IDs are filtered out
   - Error toast appears
   - Only valid, authorized programs shown

**As a Non-Reviewer:**

1. Try accessing `/community/{slug}/admin/milestones-report` directly
2. Verify access is denied (not authorized message)

**As an Admin:**

1. Access milestone report page
2. Verify you see ALL programs in dropdown (no filtering)
3. Invalid program IDs still filtered for data integrity

**Expected:**
- ✅ Milestone reviewers: Filtered dropdown
- ✅ Non-reviewers: Access denied
- ✅ Admins: Full access to all programs
- ✅ URL tampering prevented with validation
- ✅ User-friendly error messages

### 4. Milestone Verification Permissions

**As a Milestone Reviewer:**

1. Navigate to project milestone review page
2. Verify you can:
   - See "Verify" buttons on unverified milestones
   - See "Sync" buttons on off-chain verified milestones
   - Complete milestone verification
3. Test verification flow works correctly

**As a Program Reviewer (not milestone reviewer):**

1. Try accessing milestone review page
2. Verify you DON'T see:
   - Verify buttons
   - Sync buttons
   - Complete buttons

**Expected:**
- ✅ Milestone reviewers can verify/sync
- ✅ Program reviewers cannot verify/sync (view only)
- ✅ Admins and contract owners can verify/sync
- ✅ Clear permission boundaries

### 5. Edge Cases & Validation

- **Invalid wallet address:** Shows "Invalid wallet address format"
- **Invalid email:** Shows "Invalid email format"
- **Invalid telegram:** Shows proper format message
- **Duplicate reviewer:** Shows "This member already exists" (409 error)
- **Network error:** Shows retry-friendly error message
- **User is both program AND milestone reviewer:** Both roles shown with badges
- **Malformed member ID in URL:** Caught and handled gracefully
- **XSS attempt in inputs:** Sanitized before processing
- **Empty community:** Shows proper empty state
- **No programs assigned:** "View Milestones" button doesn't appear

## 📐 Architecture

```mermaid
graph TB
    subgraph "Frontend Components"
        A[My Reviews Page] --> B[Community Cards]
        B --> C[View Programs Button]
        B --> D[View Milestones Button]
        D --> E[Milestone Report Page]
        E --> F[Access Control Layer]
        F --> G[URL Validator]

        H[Reviewer Management Tab] --> I[Role Toggle]
        I --> J[Program Reviewers List]
        I --> K[Milestone Reviewers List]
        H --> L[Member ID Parser]
    end

    subgraph "Validation Layer"
        M[validators.ts] --> N[validateWalletAddress]
        M --> O[validateProgramIdentifier]
        M --> P[parseReviewerMemberId]
        M --> Q[validateReviewerData]
    end

    subgraph "Hooks & Services"
        R[useReviewerPrograms] --> S[PermissionsService]
        T[useMilestoneReviewers] --> U[MilestoneReviewersService]

        S --> V[GET my-reviewer-programs]
        U --> W[GET milestone-reviewers]
        U --> X[POST milestone-reviewers]
        U --> Y[DELETE milestone-reviewers]
    end

    subgraph "Backend API"
        V --> Z[gap-indexer]
        W --> Z
        X --> Z
        Y --> Z
    end

    subgraph "Type System"
        AA[types.ts] --> AB[ReviewerRole]
        AA --> AC[RoleMember]
        AD[helpers.ts] --> AE[getMemberRole]
        AD --> AF[getRoleLabel]
    end

    A -.uses.-> R
    E -.uses.-> R
    E -.uses.-> M
    H -.uses.-> T
    H -.uses.-> M
    L -.uses.-> P
    G -.uses.-> O
    T -.uses.-> Q
    J -.uses.-> AE
    K -.uses.-> AF

    style D fill:#a8dadc
    style E fill:#a8dadc
    style K fill:#a8dadc
    style T fill:#a8dadc
    style U fill:#a8dadc
    style M fill:#f4a261
    style AA fill:#e76f51
    style AD fill:#e76f51
```

### Access Control
- ✅ Permission checks on milestone report page
- ✅ Permission checks on milestone review/verification
- ✅ Program dropdown filtering based on user role
- ✅ URL parameter validation to prevent tampering
- ✅ Backend validates all operations (frontend is UI layer only)

### Input Validation
- ✅ Wallet address format validation (0x + 40 hex chars)
- ✅ Email format validation
- ✅ Telegram handle validation (5-32 alphanumeric)
- ✅ Program ID format validation
- ✅ Chain ID validation (positive integer)
- ✅ Combined identifier validation (programId_chainID)
- ✅ XSS prevention through input sanitization

### Data Privacy
- No sensitive data exposed in reviewer lists
- User profiles fetched securely via authenticated endpoints
- Proper authentication checks throughout
- Error messages don't leak sensitive information

### Type Safety
- No `any` types that could hide bugs
- Proper error type guards
- Type-safe API error handling
- Runtime validation matches TypeScript types

### Code Quality
- [x] Follows coding standards
- [x] No console.log statements (except development debugging)
- [x] Proper error handling with user-friendly messages
- [x] Type safety maintained (no `any` types)
- [x] Input validation and sanitization
- [x] Memoization for performance

### Testing
- [ ] Unit tests added/updated (manual testing completed)
- [x] Integration tests pass (manual verification)
- [ ] E2E tests if applicable
- [x] Manual testing completed across all scenarios
- [x] Edge cases tested

### Documentation
- [x] Code comments added for complex logic
- [x] README updated if needed
- [x] API integration documented
- [x] Migration guide not needed (additive changes)
- [x] Comprehensive PR description

### Security
- [x] Input validation implemented
- [x] Access control verified
- [x] No XSS vulnerabilities
- [x] Type safety ensures runtime safety
- [x] Error handling doesn't leak information

### Review
- [x] Self-review completed
- [x] PR title is descriptive
- [x] Commits are atomic
- [x] Ready for review

## 🔗 Related Issues

- Relates to milestone reviewer feature request
- Depends on backend PR: [gap-indexer#646](https://github.com/show-karma/gap-indexer/pull/646)

### Key Areas to Review

1. **Security & Validation** (`utilities/validators.ts`):
   - Verify validation logic is sound
   - Check regex patterns are correct
   - Ensure all edge cases are handled

2. **Permission Logic** (`components/Pages/Admin/ReportMilestonePage.tsx`, `MilestonesReview/index.tsx`):
   - Verify access control logic is sound
   - Check that admins bypass filtering correctly
   - Ensure milestone reviewers have correct permissions

3. **Type Safety** (all files):
   - Confirm no `any` types remain
   - Verify type guards are correct
   - Check error handling is type-safe

4. **UX Flow** (`app/my-reviews/page.tsx`):
   - Validate button placement and styling
   - Check URL construction for milestone links
   - Verify role badges display correctly

5. **Service Layer** (`services/milestone-reviewers.service.ts`, `services/program-reviewers.service.ts`):
   - Review API endpoint structure
   - Validate error handling
   - Check usage of shared validators

6. **Helper Utilities** (`components/Generic/RoleManagement/helpers.ts`):
   - Verify role extraction logic
   - Check label generation
   - Ensure reusability

### Testing Focus

- Test URL tampering scenarios thoroughly
- Verify permission boundaries are respected
- Test all validation error messages
- Check confirmation dialogs show correct information
- Validate error handling for API failures

### Known Considerations

- Build warnings (console.log, img tags, etc.) are pre-existing and not introduced by this PR
- TypeScript strict mode passes successfully
- All new code follows project conventions

## 🚀 Post-Merge Actions

- [ ] Deploy backend PR first: [gap-indexer#646](https://github.com/show-karma/gap-indexer/pull/646)
- [ ] Deploy to staging
- [ ] Manual smoke testing on staging
- [ ] Monitor for errors in Sentry
- [ ] Notify team of new milestone reviewer feature
- [ ] Update user documentation if needed

---

## 📊 Code Stats

**13 files changed:**
- **1,256 insertions**
- **205 deletions**
- **Net: +1,051 lines**

**New files:**
- `utilities/validators.ts` (274 lines) - Shared validation utilities
- `components/Generic/RoleManagement/helpers.ts` (38 lines) - Role helper functions
- `hooks/useMilestoneReviewers.ts` (95 lines) - Milestone reviewer hook
- `services/milestone-reviewers.service.ts` (250 lines) - Milestone reviewer API service

**Major refactors:**
- `components/FundingPlatform/QuestionBuilder/ReviewerManagementTab.tsx` - Dual-role management
- `components/Generic/RoleManagement/RoleManagementTab.tsx` - Type safety + UX improvements
- `components/Pages/Admin/ReportMilestonePage.tsx` - Access control + validation

---

**Note:** This PR requires the corresponding backend changes to be deployed first. Ensure gap-indexer PR is merged and deployed before deploying this frontend PR.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1211667339736200

### CODE REVIEW

File: components/FundingPlatform/QuestionBuilder/ReviewerManagementTab.tsx [hunk 1]
```diff
@@ -69,39 +95,47 @@ export const ReviewerManagementTab: React.FC<ReviewerManagementTabProps> = ({
       });
     },
     onSuccess: async () => {
-      await queryClient.invalidateQueries({ queryKey: ["program-reviewers", programId, chainID] });
-      toast.success("Reviewer added successfully");
+      await queryClient.invalidateQueries({
+        queryKey: ["program-reviewers", programId, chainID],
```
Comment (HadiKhai) [inline components/FundingPlatform/QuestionBuilder/ReviewerManagementTab.tsx]: use builder keys or have a queryKeys files nd put all the keys there so that "program-reviewers" with the params are defined once in the code
